staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config.go

```diff
@@ -44,8 +49,88 @@ const (
 	secretboxTransformerPrefixV1 = "k8s:enc:secretbox:v1:"
 	kmsTransformerPrefixV1       = "k8s:enc:kms:v1:"
 	kmsPluginConnectionTimeout   = 3 * time.Second
+	kmsPluginHealthzRate         = rate.Limit(1)
+	kmsPluginHealthzBurst        = 1
 )
 
+// KMSPluginProbe checks health status of KMS-Plugins.
+type KMSPluginProbe struct {
+	name string
+	envelope.Service
+	*rate.Limiter
+}
+
+// GetKMSPluginProbes extracts KMSPluginProbes from the EncryptionConfig.
+func GetKMSPluginProbes(filepath string) ([]*KMSPluginProbe, error) {
+	f, err := os.Open(filepath)
+	if err != nil {
+		return nil, fmt.Errorf("error opening encryption provider configuration file %q: %v", filepath, err)
+	}
+	defer f.Close()
+	return getKMSPluginProbes(f)
+}
+
+func getKMSPluginProbes(reader io.Reader) ([]*KMSPluginProbe, error) {
+	var result []*KMSPluginProbe
+
+	configFileContents, err := ioutil.ReadAll(reader)
+	if err != nil {
+		return result, fmt.Errorf("could not read content of encryption provider configuration: %v", err)
+	}
+
+	config, err := loadConfig(configFileContents)
+	if err != nil {
+		return result, fmt.Errorf("error while parsing encrypiton provider configuration: %v", err)
+	}
+
+	for _, r := range config.Resources {
+		for _, p := range r.Providers {
+			if p.KMS != nil && p.KMS.InstallHealthz {
```

Comment:
it's been a very long while since I looked in here, but I thought the API was "bytes in, bytes out".  If that happens, aren't you healthy?  If so, why is this optional and why are we building a separate client for it?

Reply:
> it's been a very long while since I looked in here, but I thought the API was "bytes in, bytes out". If that happens, aren't you healthy? If so, why is this optional and why are we building a separate client for it?

Not so much against the separate construction as the factorization that I expect looks nearly identical.

Reply:
I agree, there is no reason why this should be optional - now installing healthz check per KMS Provider.

Regarding the multiple connections, I think having multiple connections (with different dial options) may become useful in the future. For example, we may decide to FailFast in healthz. Having said that, I am open to refactoring this to a single connection.